src/plugins/clips-executive/clips/domain.clp

```diff
@@ -178,6 +178,260 @@
+=>
+  (modify ?parent (is-satisfied FALSE))
+)
+
+(defrule domain-check-if-conjunctive-formula-is-satisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied FALSE))
+  (pddl-formula (id ?base) (type conjunction))
+
+  (not (grounded-pddl-formula (part-of ?id) (is-satisfied FALSE) (id ?child-id&:(neq ?id ?child-id))))
+  (not (grounded-pddl-predicate (part-of ?id) (is-satisfied FALSE)))
+=>
+  (modify ?parent (is-satisfied TRUE))
+)
+
+(defrule domain-check-if-conjunctive-formula-is-unsatisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied TRUE))
+  (pddl-formula (id ?base) (type conjunction))
+
+  (or (grounded-pddl-formula (is-satisfied FALSE) (part-of ?id) (id ?child-id&:(neq ?id ?child-id)))
+      (grounded-pddl-predicate (is-satisfied FALSE) (part-of ?id))
+  )
+=>
+  (modify ?parent (is-satisfied FALSE))
+)
+
+(defrule domain-check-if-disjunctive-formula-is-satisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied FALSE))
+  (pddl-formula (id ?base) (type disjunction))
+
+  (or (grounded-pddl-formula (is-satisfied TRUE) (part-of ?id) (id ?child-id&:(neq ?id ?child-id)))
+      (grounded-pddl-predicate (is-satisfied TRUE) (part-of ?id))
+  )
+ =>
+  (modify ?parent (is-satisfied TRUE))
+)
+
+(defrule domain-check-if-disjunctive-formula-is-unsatisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  
+  ?parent <- (grounded-pddl-formula (id ?id) (formula-id ?base) (is-satisfied TRUE))
+  (pddl-formula (id ?base) (type disjunction))
+
+  (not (or (grounded-pddl-formula (is-satisfied TRUE) (part-of ?id) (id ?child-id&:(neq ?id ?child-id)))
+      (grounded-pddl-predicate (is-satisfied TRUE) (part-of ?id))
+  ))
+=>
+  (modify ?parent (is-satisfied FALSE))
+)
+
+(defrule domain-check-if-predicate-is-satisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  ?predicate <- (grounded-pddl-predicate (param-values $?params) 
+                                         (is-satisfied FALSE) 
+                                         (predicate-id ?id))
+  ?base <- (pddl-predicate (id ?id) (equality FALSE) (predicate ?pred))                  
+  (domain-fact (name ?pred) (param-values $?params))
+=>
+  (modify ?predicate (is-satisfied TRUE))
+)
+
+(defrule domain-check-if-predicate-equality-is-satisfied
+  (declare (salience ?*SALIENCE-DOMAIN-CHECK*))
+  ?predicate <- (grounded-pddl-predicate 
+                  (is-satisfied ?is-sat) 
+                  (predicate-id ?id)
+                  (param-values $?params& :
+                    (and (= (length$ ?params) 2)
+                       (neq ?is-sat (eq (nth$ 1 ?params) (nth$ 2 ?params))))
+                  ))
+  ?base <- (pddl-predicate (id ?id) (equality TRUE))
+=>
+  (modify ?predicate (is-satisfied (eq (nth$ 1 ?params) (nth$ 2 ?params))))
+)
```

Comment:
I expect that this will create an endless loop. Better check for equality directly on the LHS:
```
?predicate <- (grounded-pddl-predicate (is-satisfied FALSE) (param-values ?p1 ?p2&:(eq ?p1 ?p2)) ...)
=>
(modify ?predicate (is-satisfied TRUE))
```

Reply:
Also, this will silently fail if the number of parameters is not two. That should give an error message.